api/envoy/extensions/filters/listener/proxy_protocol/v3/proxy_protocol.proto

```diff
@@ -86,3 +88,12 @@ message ProxyProtocol {
   //   The filter treats this request as if it did not have any PROXY protocol information.
   repeated config.core.v3.ProxyProtocolConfig.Version disallowed_versions = 4;
 }
+
+message TlvsMetadata {
+  // Typed metadata for :ref:`Proxy protocol filter <envoy_v3_api_msg_extensions.filters.listener.proxy_protocol.v3.ProxyProtocol>`, that represents map of TLVs.
+  // Each entry in the map consists of a key (TLV type) and value (TLV value in bytes).
+  // When runtime flag ``envoy.reloadable_features.use_typed_metadata_in_proxy_protocol_listener`` is enabled,
+  // :ref:`Proxy protocol filter <envoy_v3_api_msg_extensions.filters.listener.proxy_protocol.v3.ProxyProtocol>`
+  // will set typed metadata instead of regular metadata. By default filter will use typed metadata.
+  map<string, google.protobuf.BytesValue> typed_metadata = 1;
```

Comment:
The type/key may not be a utf-8 string; it is a single byte (any byte). So a `string` for the key is incorrect.

Also, there's nothing in the spec specifically prohibiting using a type/key more than once, so maybe this should be a `repeated` instead of `map`?

Reply:
Also, why use `BytesValue` instead of just `bytes`?

Reply:
@ggreenway no particular reason for using `BytesValue` over `bytes`. We seem to use both across the codebase.

Reply:
@ggreenway in the above map key is not a single byte, it is so called [`ProxyProtocol.Rule` key](https://github.com/envoyproxy/envoy/blob/8d4cf00e2980aeba11b9b74230a53f92086419bc/api/envoy/extensions/filters/listener/proxy_protocol/v3/proxy_protocol.proto#L25) coming from the Envoy config, while the value is entire tlv in raw bytes ({type, lenght_hi, lenght_lo, value}). I have replicated proxy protocol filter untyped metadata [structure](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc#L535). Dynamic metadata (aka `envoy::config::core::v3::Metadata`) contains untyped filters metadata in form of:
```
map<string, google.protobuf.Struct> filter_metadata
```
Where proxy protocol filter currently sets an entry in this map as:
```
"envoy.filters.listener.proxy_protocol", ProtobufWkt::Struct
```
Where the `Struct` object looks smth like this:
```
field 0 (name, std:string: "PP2 vpc id"), value 0 (bytes {0xea, 0x00, 0x16, 0x01 ...})
field 1 (name, std:string: "PP2 type authority"), value 1 (bytes {0xea, 0x00, 0x13, 0x02 ...})
```

Reply:
ok, i see the proto docs are misleading and refer to key as Tlv type. Will fix.

Reply:
Got it; I was thinking this string was the [`tlv_type`](https://github.com/envoyproxy/envoy/blob/8d4cf00e2980aeba11b9b74230a53f92086419bc/api/envoy/extensions/filters/listener/proxy_protocol/v3/proxy_protocol.proto#L38). String type makes sense as you have it.